### PR TITLE
tests: refactor lxd launcher tests

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

The unit tests in tests/unit/lxd/test_launcher.py were designed to work with a single instance.  They use fixtures that do not distinguish which calls are made to which instance.

This is preventing me from testing https://github.com/canonical/craft-providers/pull/177, where I need the tests to distinguish between and instance and a base instance.

This PR refactors the fixtures and does a little cleanup.

(CRAFT-1478)

### CODE REVIEW

File: tests/unit/lxd/test_launcher.py [hunk 1]
```diff
@@ -24,15 +25,15 @@
...
+    mock_base = Mock(spec=Base)
     mock_base.compatibility_tag = "mock-compat-tag-v100"
     mock_base.get_command_environment.return_value = {"foo": "bar"}
     yield mock_base
 
 
 @pytest.fixture
 def mock_lxc():
-    with mock.patch(
+    with patch(
```
Comment (sergiusens) [inline tests/unit/lxd/test_launcher.py:36]: maybe move this to use mocker?

Comment (cmatsuoka) [reply tests/unit/lxd/test_launcher.py:36]: +1 for the pytest mocker fixture

File: tests/unit/lxd/test_launcher.py [hunk 2]
```diff
@@ -41,22 +42,35 @@ def mock_lxc():
...
         spec=lxd.LXDInstance,
-    ) as mock_instance:
-        mock_instance.return_value.name = "test-instance-$"
-        # the name has an invalid character, so the instance_name will be different
-        mock_instance.return_value.instance_name = "test-instance-fa2d407652a1c51f6019"
-        mock_instance.return_value.project = "test-project"
-        mock_instance.return_value.remote = "test-remote"
-        yield mock_instance
-
+        # single element list appears accidental but side_effect must be an interable.
```
Comment (cmatsuoka) [inline tests/unit/lxd/test_launcher.py:65]: Is this a typo for iterable?

File: tests/unit/lxd/test_launcher.py [hunk 3]
```diff
@@ -41,22 +42,35 @@ def mock_lxc():
...
 @pytest.fixture
-def mock_lxd_instance():
-    with mock.patch(
+def fake_instance():
+    """Returns a fake LXD Instance"""
+    instance = MagicMock()
+    # the name has an invalid character to ensure the instance_name will be different
+    # so the two are not conflated in the unit tests
+    instance.name = "test-instance-$"
+    instance.instance_name = "test-instance-fa2d407652a1c51f6019"
```
Comment (cmatsuoka) [inline tests/unit/lxd/test_launcher.py:51]: I'm not familiar with the internal structure of instances but `instance.name` and `instance.instance_name` sound a bit confusing. Just mentioning, I don't think there's much we can do about that here in tests.

**Review by cmatsuoka — APPROVED** [review]
Thanks for updating. The changes are a bit hard to follow but the overall concept seems good. Pointed a possible typo in a comment.

File: tests/unit/lxd/test_launcher.py (hunk 3 above)
Comment (mr-cal) [reply tests/unit/lxd/test_launcher.py:48]: I agree with you, it's a confusing subtlety that has arisen from craft-provider's evolving without breaking changes.

File: tests/unit/lxd/test_launcher.py (hunk 2 above)
Comment (mr-cal) [reply tests/unit/lxd/test_launcher.py]: Yes, thanks.  Interesting that codespell doesn't catch it.

File: tests/unit/lxd/test_launcher.py (hunk 1 above)
Comment (mr-cal) [reply tests/unit/lxd/test_launcher.py]: Added.
